### ISSUE TITLE
Problem with Scroll

### ISSUE THREAD

**Issue by librecourage** [elastic/elasticsearch-net#326]
With the version 0.11.2, i can't retrieve results via scrolling

I launch a query. The response is : 
{"_scroll_id":"c2Nhbjs1Ozc3NjgyOjYwZFRpNmJwVGZPdUpRNkh2cmRBSmc7Nzc2ODU6NjBkVGk2YnBUZk91SlE2SHZyZEFKZzs3NzY4NDo2MGRUaTZicFRmT3VKUTZIdnJkQUpnOzc3NjgxOjYwZFRpNmJwVGZPdUpRNkh2cmRBSmc7Nzc2ODM6NjBkVGk2YnBUZk91SlE2SHZyZEFKZzsxO3RvdGFsX2hpdHM6Mjs=","took":63,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":2,"max_score":0.0,"hits":[]}}

When i want to see the results, Nest sends this message : 
c2Nhbjs1Ozc3NjgyOjYwZFRpNmJwVGZPdUpRNkh2cmRBSmc7Nzc2ODU6NjBkVGk2YnBUZk91SlE2SHZyZEFKZzs3NzY4NDo2MGRUaTZicFRmT3VKUTZIdnJkQUpnOzc3NjgxOjYwZFRpNmJwVGZPdUpRNkh2cmRBSmc7Nzc2ODM6NjBkVGk2YnBUZk91SlE2SHZyZEFKZzsxO3RvdGFsX2hpdHM6Mjs%3D

and i receive this response

{"error":"ElasticSearchIllegalArgumentException[Failed to decode scrollId]; nested: IOException[Bad Base64 input character decimal 37 in array position 223]; ","status":400}

**Comment by royborgiani** [comment]
I´m having this same problem with scroll, on ElasticSearch 1.7.1 

{
   "error": "ElasticsearchIllegalArgumentException[Failed to decode scrollId]; nested: IOException[Bad Base64 input character decimal 123 in array position 0]; ",
   "status": 400
}

And I need to delete some items from a index with type boolean (true)

How can I solve this problem?

**Comment by russcam** [comment]
@royborgiani, what version of NEST are you using with Elasticsearch 1.7.1?

Also, do you have a concise example of what you're doing?

**Comment by royborgiani** [comment]
How I GET the NEST version?

I do the following, in the sense, I Execute the command below:

GET /twitter/tweet/_search?scroll=1m
{
    "query": {
        "match" : {
            "type" : "false"
        }
    }
}

and I get this return:

{
   "_scroll_id": "cXVlcnlUaGVuRmV0Y2g7NTs5ODpIUy1RLVFvbVNBcU1EX0lnNkZKNUdBOzk5OkhTLVEtUW9tU0FxTURfSWc2Rko1R0E7MTAwOkhTLVEtUW9tU0FxTURfSWc2Rko1R0E7MTAxOkhTLVEtUW9tU0FxTURfSWc2Rko1R0E7MTAyOkhTLVEtUW9tU0FxTURfSWc2Rko1R0E7MDs=",
   "took": 9,
   "timed_out": false,
   "_shards": {
      "total": 5,
      "successful": 5,
      "failed": 0
   },
   "hits": {
      "total": 2,
      "max_score": 0.30685282,
      "hits": [with some hits that is my clause to limit the returns]

Then I: get the _scroll_id insert with the following command:

DELETE /_search/scroll -d 'ccXVlcnlUaGVuRmV0Y2g7NTs1OTg6SFMtUS1Rb21TQXFNRF9JZzZGSjVHQTs1OTk6SFMtUS1Rb21TQXFNRF9JZzZGSjVHQTs2MDA6SFMtUS1Rb21TQXFNRF9JZzZGSjVHQTs2MDE6SFMtUS1Rb21TQXFNRF9JZzZGSjVHQTs2MDI6SFMtUS1Rb21TQXFNRF9JZzZGSjVHQTswOw=='

So, I get that error:

{
   "error": "ElasticsearchIllegalArgumentException[Failed to decode scrollId]; nested: IOException[Bad Base64 input character decimal 47 in array position 7]; ",
   "status": 400
}

:(

**Comment by russcam** [comment]
Check the nuget package version of NEST that you are using in `packages.config` (or `project.json` if using xproj). If you're using Elasticsearch 1.7.1, then [the latest NEST version that is compatible is NEST 1.8.0](https://www.nuget.org/packages/NEST/1.8.0).

In your example above, where is the `scroll_id` value that you send to delete/clear the scroll coming from? It's not the same as the `scroll_id` that is returned from the search?

**Comment by royborgiani** [comment]
in the example no, but, i´m using the same scroll_id like below:

GET /twitter/tweet/_search?scroll=1m
{
    "query": {
        "match" : {
            "type" : "false"
        }
    }
}

GET /_search/scroll
{
    "scroll" : "1m", 
    "scroll_id" : "cXVlcnlUaGVuRmV0Y2g7NTs2ODg2NjpIUy1RLVFvbVNBcU1EX0lnNkZKNUdBOzY4ODY3OkhTLVEtUW9tU0FxTURfSWc2Rko1R0E7Njg4Njg6SFMtUS1Rb21TQXFNRF9JZzZGSjVHQTs2ODg2OTpIUy1RLVFvbVNBcU1EX0lnNkZKNUdBOzY4ODcwOkhTLVEtUW9tU0FxTURfSWc2Rko1R0E7MDs=" 
}

and the return of the second command is:

{
   "error": "ElasticsearchIllegalArgumentException[Failed to decode scrollId]; nested: IOException[Bad Base64 input character decimal 123 in array position 0]; ",
   "status": 400
}

Finally, the same result as on the last comment before...i see this error, on delete or get...

**Comment by asd-and-Rizzo** [comment]
I used NEST 1.6.1 against ES 1.7 and things where fine (next code is part of batching enumerator):
```
class ScrollEnumerator<TCollection, T> : IEnumerator<TCollection>
            where TCollection : IEnumerable<T>
...
MoveNext()
...
                if (string.IsNullOrEmpty(this.scrollId))
                {
                    //// NOTE: use const from NEST 2
                    const string SortSpecialFieldDocumentIndexOrder = "_doc";
                        searchResult = this.client.Search<T>(selector => selector
                            .Index(...)
                            .Types(typeof(T))
                            //// NOTE: will return shards count * maxWindowSize
                            .Size(this.maxWindowSize)
                            .Source(sr => sr.Include(this.include))
                            .Scroll(..600 seconds...)
                            .Sort(x => x.OnField(SortSpecialFieldDocumentIndexOrder))
                            .Query(query => query.Filtered(f => f.Filter(fi => fi.Bool(b => b.Must(this.filters))))));
                }
                else
                {
                        searchResult = this.client.Scroll<T>(..600 seconds..., this.scrollId);
                 }
                this.scrollId = searchResult.ScrollId;
```

I use NEST 2.1.1 against ES 2.1 as next and get error with scroll (first call passes, second batch fails):
```
                if (string.IsNullOrEmpty(this.scrollId))
                {
                        var searchDescriptor = new SearchDescriptor<T>()
                            .Index(...)
                            .Type<T>()
                            //// NOTE: will return shards count * maxWindowSize
                            .Size(this.maxWindowSize)
                            .Source(sr => sr.Include(fs => fs.Fields(this.include)))
                            .Scroll(..600 seconds...)
                            .Sort(x => x.Ascending(SortSpecialField.DocumentIndexOrder))
                            .QueryExactNonScored(this.filters);
#if DEBUG
                        Debug.WriteLine(this.client.ToRawRequest(searchDescriptor));
#endif
                        searchResult = this.client.Search<T>(searchDescriptor);
                }
                else
                {
                        searchResult = this.client.Scroll<T>(..600 seconds..., this.scrollId);
                }
                this.scrollId = searchResult.ScrollId;


        public static SearchDescriptor<T> QueryExactNonScored<T>(this SearchDescriptor<T> self, QueryContainer exactNonScoredQuery) where T : class
        {
#pragma warning disable 618 // Type or member is obsolete
            return self.Query(query => query.Filtered(f => f.Filter(fi => fi.Bool(b => b.Must(exactNonScoredQuery)))));
#pragma warning restore 618 // Type or member is obsolete
        }

       public static string ToRawRequest<T>(this IElasticClient self, SearchDescriptor<T> searchDescriptor) where T : class
        {
            using (var output = new MemoryStream())
            {
                self.Serializer.Serialize(searchDescriptor, output);
                output.Position = 0;
                var rawQuery = new StreamReader(output).ReadToEnd();
                return rawQuery;
            }
        }
```

Error is next:
```
Response
HTTP/1.1 400 Bad Request
Auth-Scheme: BASIC
User: caymanes_write
Content-Type: application/json; charset=UTF-8
Content-Length: 184

{
  "error" : "ElasticsearchIllegalArgumentException[Failed to decode scrollId]; nested: IOException[Bad Base64 input character decimal 123 in array position 0]; ",
  "status" : 400
}
```

**Comment by asd-and-Rizzo** [comment]
NEST 2.1 request:
```
POST: /_search/scroll?pretty=true
{
  "scroll": "600s",
  "scroll_id": "cXVlcnlBbmRGZXRjaDsxOzQ2NDE0Njg6MHhVOHlQeXpRMENmSkIwMDQ0NEdGdzswOw=="
}
```

NEST 1.6.1 request:
```
POST: : .../_search/scroll?scroll=600s&pretty=true, 
	Request: cXVlcnlBbmRGZXRjaDsxOzQ2NDE0NDE6MHhVOHlQeXpRMENmSkIwMDQ0NEdGdzswOw== 
```

As it is visible newer version send JSON, while older send base64 string and query string parameters.

**Comment by kurokuriboh** [comment]
Is this issue still open? I still have the issue for Nest 2.3.1

**Comment by russcam** [comment]
@kurokuriboh looking into it

**Comment by kurokuriboh** [comment]
I actually found out the issue. Aws still uses elasticsearch 1.5 and the bug is not fixed until later version. The issue can be closed I believe. Sorry for that

**Comment by russcam** [comment]
I'll close it after we're happy that it's not a wider issue.

By the way, you should use the latest NEST 1.x release with Elasticsearch 1.5; NEST 2.x is not supported as being compatible. It _may_ work for the most part, but there are breaking changes between Elasticsearch 1.x and 2.x that are reflected in changes to NEST, meaning that some things will not work at all.

**Comment by Mpdreamz** [comment]
Just validated this works as expected with `Elasticsearch 1.x` and `Nest 1.x`. The original fix was to **not** encode the traling `=` at the last position the fix for which went into Nest 0.12. 

If you are seeing a problem at position 0, you are most likely sending json to Elasticsearch 1.x using a 2.x client. The scroll API only supports json since Elasticsearch 2.0.0

**Comment by geremyCohen** [comment]
FTR, the equivalent CURL commands against ES 1.5 are:

```
curl -XGET "https://localhost/twitter/tweets/_search?scroll=2m"
curl -XGET "https://localhost/_search/scroll?scroll=2m" -d 'SCROLL_ID_RESPONSE_VALUE_FROM_LAST_REQUEST'
```
